SINGA-510 Distributed Training Time Profiling

chrishkchris opened the pull request:
(Updated on 2 June 2020)

I am working on the distributed training time profiling

Done:
1. Moved the cuda streams from `Communicator` to `Context`
2. Time Profiling for Commmunicator buffered operations
3. Break down some communication kernel using different streams to exclude the waiting time in profiling
4. Added verbosity level 3 for showing distributed operators only
5. Rewrote the backward time (verbosity=1) to adapt for distributed training

Current Result tested in panda 18 (eight GPUs) is:
```
root@e8a757397ca3:~/dcsysh/singa/examples/cnn# mpiexec -np 8 python3 benchmark.py --dist --verbosity 3
100%|##########| 100/100 [00:18<00:00,  5.51it/s]Throughput = 1389.5021244306654 per second
TotalTime=18.42386531829834
Total=0.1842386531829834
Time Profiling:
OP_ID242. Dist_c1_fusedSynch_filling : 1.04998e-05 sec
OP_ID246. Dist_c1_fusedSynch_filling : 0.000351449 sec
OP_ID251. Dist_c1_fusedSynch_filling : 1.8154e-05 sec
OP_ID253. Dist_c1_fusedSynch_filling : 9.87106e-06 sec
OP_ID257. Dist_c1_fusedSynch_filling : 9.02927e-05 sec
OP_ID259. Dist_s_fusedSynch_allreduce : 0.0117456 sec
OP_ID261. Dist_c1_fusedSynch_copyBackToTensor : 0.000577933 sec
OP_ID265. Dist_c1_fusedSynch_filling : 0.000192158 sec
OP_ID267. Dist_c1_fusedSynch_filling : 0.000229652 sec
OP_ID271. Dist_s_synch_allreduce : 0.00892664 sec
OP_ID275. Dist_c1_fusedSynch_filling : 0.000302222 sec
OP_ID277. Dist_c1_fusedSynch_filling : 0.000154409 sec
OP_ID282. Dist_c1_fusedSynch_filling : 0.000156862 sec
OP_ID286. Dist_c1_fusedSynch_filling : 0.000137269 sec
OP_ID288. Dist_c1_fusedSynch_filling : 9.43078e-05 sec
OP_ID292. Dist_c1_fusedSynch_filling : 0.000118389 sec
OP_ID294. Dist_s_fusedSynch_allreduce : 0.00832119 sec
OP_ID296. Dist_c1_fusedSynch_copyBackToTensor : 0.00105477 sec
OP_ID300. Dist_c1_fusedSynch_filling : 0.000104583 sec
OP_ID302. Dist_c1_fusedSynch_filling : 0.000129235 sec
OP_ID306. Dist_s_synch_allreduce : 0.00882223 sec
OP_ID310. Dist_c1_fusedSynch_filling : 0.000117987 sec
OP_ID312. Dist_c1_fusedSynch_filling : 0.000130123 sec
OP_ID317. Dist_c1_fusedSynch_filling : 0.000157646 sec
OP_ID321. Dist_c1_fusedSynch_filling : 0.000122986 sec
OP_ID323. Dist_c1_fusedSynch_filling : 0.000122919 sec
OP_ID327. Dist_c1_fusedSynch_filling : 0.0001817 sec
OP_ID329. Dist_s_fusedSynch_allreduce : 0.0122809 sec
OP_ID331. Dist_c1_fusedSynch_copyBackToTensor : 0.00135137 sec
OP_ID334. Dist_c1_fusedSynch_filling : 0.000166807 sec
OP_ID336. Dist_c1_fusedSynch_filling : 0.000215435 sec
OP_ID340. Dist_c1_fusedSynch_filling : 0.000181097 sec
OP_ID344. Dist_c1_fusedSynch_filling : 0.000208481 sec
OP_ID346. Dist_c1_fusedSynch_filling : 0.000172447 sec
OP_ID350. Dist_s_synch_allreduce : 0.00910959 sec
OP_ID354. Dist_c1_fusedSynch_filling : 0.000155029 sec
OP_ID356. Dist_c1_fusedSynch_filling : 0.000220342 sec
OP_ID361. Dist_c1_fusedSynch_filling : 0.000218153 sec
OP_ID365. Dist_c1_fusedSynch_filling : 0.000170545 sec
OP_ID367. Dist_c1_fusedSynch_filling : 0.000175028 sec
OP_ID371. Dist_c1_fusedSynch_filling : 0.000228723 sec
OP_ID375. Dist_c1_fusedSynch_filling : 0.000209001 sec
OP_ID377. Dist_c1_fusedSynch_filling : 0.000149317 sec
OP_ID381. Dist_c1_fusedSynch_filling : 0.000178857 sec
OP_ID383. Dist_s_fusedSynch_allreduce : 0.00957942 sec
OP_ID385. Dist_c1_fusedSynch_copyBackToTensor : 0.00111657 sec
OP_ID389. Dist_c1_fusedSynch_filling : 0.000113132 sec
OP_ID391. Dist_c1_fusedSynch_filling : 6.6741e-05 sec
OP_ID396. Dist_c1_fusedSynch_filling : 7.78491e-05 sec
OP_ID400. Dist_c1_fusedSynch_filling : 8.09889e-05 sec
OP_ID402. Dist_c1_fusedSynch_filling : 7.4416e-05 sec
OP_ID406. Dist_c1_fusedSynch_filling : 7.73628e-05 sec
OP_ID410. Dist_c1_fusedSynch_filling : 7.95015e-05 sec
OP_ID412. Dist_c1_fusedSynch_filling : 0.000118768 sec
OP_ID416. Dist_c1_fusedSynch_filling : 0.000113379 sec
OP_ID420. Dist_c1_fusedSynch_filling : 0.000105051 sec
OP_ID422. Dist_c1_fusedSynch_filling : 9.89747e-05 sec
OP_ID427. Dist_c1_fusedSynch_filling : 0.000113316 sec
OP_ID431. Dist_c1_fusedSynch_filling : 8.11896e-05 sec
OP_ID433. Dist_c1_fusedSynch_filling : 0.000131457 sec
OP_ID437. Dist_c1_fusedSynch_filling : 0.000114805 sec
OP_ID441. Dist_c1_fusedSynch_filling : 0.000111409 sec
OP_ID443. Dist_c1_fusedSynch_filling : 0.000117403 sec
OP_ID447. Dist_c1_fusedSynch_filling : 0.00019649 sec
OP_ID449. Dist_s_fusedSynch_allreduce : 0.00974947 sec
OP_ID451. Dist_c1_fusedSynch_copyBackToTensor : 0.00152865 sec
OP_ID455. Dist_c1_fusedSynch_filling : 3.84063e-05 sec
OP_ID457. Dist_c1_fusedSynch_filling : 5.86475e-05 sec
OP_ID462. Dist_c1_fusedSynch_filling : 6.77016e-05 sec
OP_ID466. Dist_c1_fusedSynch_filling : 0.000122606 sec
OP_ID468. Dist_c1_fusedSynch_filling : 0.000126491 sec
OP_ID472. Dist_c1_fusedSynch_filling : 0.000139088 sec
OP_ID476. Dist_c1_fusedSynch_filling : 0.000114233 sec
OP_ID478. Dist_c1_fusedSynch_filling : 0.000124881 sec
OP_ID482. Dist_c1_fusedSynch_filling : 0.00024017 sec
OP_ID486. Dist_c1_fusedSynch_filling : 0.000193712 sec
OP_ID488. Dist_c1_fusedSynch_filling : 0.000139787 sec
OP_ID493. Dist_c1_fusedSynch_filling : 0.000160467 sec
OP_ID497. Dist_c1_fusedSynch_filling : 0.000124039 sec
OP_ID499. Dist_c1_fusedSynch_filling : 0.000102468 sec
OP_ID503. Dist_c1_fusedSynch_filling : 0.000136263 sec
OP_ID507. Dist_c1_fusedSynch_filling : 0.000104248 sec
OP_ID509. Dist_c1_fusedSynch_filling : 6.39024e-05 sec
OP_ID513. Dist_c1_fusedSynch_filling : 0.000109178 sec
OP_ID515. Dist_s_fusedSynch_allreduce : 0.00908348 sec
OP_ID517. Dist_c1_fusedSynch_copyBackToTensor : 0.00133219 sec
OP_ID521. Dist_c1_fusedSynch_filling : 4.8677e-05 sec
OP_ID523. Dist_c1_fusedSynch_filling : 4.95881e-05 sec
OP_ID528. Dist_c1_fusedSynch_filling : 5.35564e-05 sec
OP_ID532. Dist_c1_fusedSynch_filling : 3.25926e-05 sec
OP_ID534. Dist_c1_fusedSynch_filling : 5.02165e-05 sec
OP_ID538. Dist_c1_fusedSynch_filling : 7.02833e-05 sec
OP_ID541. Dist_c1_fusedSynch_filling : 4.93556e-05 sec
OP_ID543. Dist_c1_fusedSynch_filling : 4.87329e-05 sec
OP_ID547. Dist_c1_fusedSynch_filling : 6.41038e-05 sec
OP_ID551. Dist_c1_fusedSynch_filling : 5.60753e-05 sec
OP_ID553. Dist_c1_fusedSynch_filling : 4.80326e-05 sec
OP_ID557. Dist_c1_fusedSynch_filling : 6.0789e-05 sec
OP_ID561. Dist_c1_fusedSynch_filling : 6.0101e-05 sec
OP_ID563. Dist_c1_fusedSynch_filling : 5.09675e-05 sec
OP_ID568. Dist_c1_fusedSynch_filling : 4.54503e-05 sec
OP_ID572. Dist_c1_fusedSynch_filling : 6.38616e-05 sec
OP_ID574. Dist_c1_fusedSynch_filling : 7.50071e-05 sec
OP_ID578. Dist_c1_fusedSynch_filling : 4.97029e-05 sec
OP_ID582. Dist_c1_fusedSynch_filling : 0.000122285 sec
OP_ID584. Dist_c1_fusedSynch_filling : 8.71332e-05 sec
OP_ID588. Dist_c1_fusedSynch_filling : 0.000107861 sec
OP_ID592. Dist_c1_fusedSynch_filling : 5.67529e-05 sec
OP_ID594. Dist_c1_fusedSynch_filling : 4.86265e-05 sec
OP_ID599. Dist_c1_fusedSynch_filling : 5.60035e-05 sec
OP_ID603. Dist_c1_fusedSynch_filling : 2.91382e-05 sec
OP_ID605. Dist_c1_fusedSynch_filling : 3.97719e-05 sec
OP_ID609. Dist_c1_fusedSynch_filling : 3.50933e-05 sec
OP_ID611. Dist_s_fusedSynch_allreduce : 0.00882422 sec
OP_ID613. Dist_c1_fusedSynch_copyBackToTensor : 0.00132458 sec
OP_ID617. Dist_c1_fusedSynch_filling : 6.99558e-05 sec
OP_ID619. Dist_c1_fusedSynch_filling : 5.8831e-05 sec
OP_ID623. Dist_c1_fusedSynch_filling : 4.7659e-05 sec
OP_ID627. Dist_c1_fusedSynch_filling : 2.93584e-05 sec
OP_ID629. Dist_c1_fusedSynch_filling : 4.52483e-05 sec
OP_ID634. Dist_c1_fusedSynch_filling : 3.98676e-05 sec
OP_ID638. Dist_c1_fusedSynch_filling : 4.58259e-05 sec
OP_ID640. Dist_c1_fusedSynch_filling : 3.44772e-05 sec
OP_ID644. Dist_c1_fusedSynch_filling : 3.9462e-05 sec
OP_ID648. Dist_c1_fusedSynch_filling : 3.13895e-05 sec
OP_ID650. Dist_c1_fusedSynch_filling : 5.74215e-05 sec
OP_ID654. Dist_c1_fusedSynch_filling : 6.24979e-05 sec
OP_ID658. Dist_c1_fusedSynch_filling : 5.90121e-05 sec
OP_ID660. Dist_c1_fusedSynch_filling : 4.94965e-05 sec
OP_ID665. Dist_c1_fusedSynch_filling : 5.71598e-05 sec
OP_ID669. Dist_c1_fusedSynch_filling : 7.19768e-05 sec
OP_ID671. Dist_c1_fusedSynch_filling : 4.30325e-05 sec
OP_ID675. Dist_c1_fusedSynch_filling : 6.52028e-05 sec
OP_ID678. Dist_c1_fusedSynch_filling : 4.42033e-05 sec
OP_ID680. Dist_c1_fusedSynch_filling : 3.94864e-05 sec
OP_ID684. Dist_c1_fusedSynch_filling : 3.92765e-05 sec
OP_ID688. Dist_c1_fusedSynch_filling : 4.83015e-05 sec
OP_ID690. Dist_c1_fusedSynch_filling : 7.2421e-05 sec
OP_ID694. Dist_c1_fusedSynch_filling : 6.2821e-05 sec
OP_ID698. Dist_c1_fusedSynch_filling : 0.000104438 sec
OP_ID700. Dist_c1_fusedSynch_filling : 4.93942e-05 sec
OP_ID705. Dist_c1_fusedSynch_filling : 8.12405e-05 sec
OP_ID709. Dist_c1_fusedSynch_filling : 5.88797e-05 sec
OP_ID711. Dist_c1_fusedSynch_filling : 7.60583e-05 sec
OP_ID715. Dist_c1_fusedSynch_filling : 5.41374e-05 sec
OP_ID719. Dist_c1_fusedSynch_filling : 4.0266e-05 sec
OP_ID721. Dist_c1_fusedSynch_filling : 4.58475e-05 sec
OP_ID725. Dist_c1_fusedSynch_filling : 4.16333e-05 sec
OP_ID729. Dist_c1_fusedSynch_filling : 3.52571e-05 sec
OP_ID731. Dist_c1_fusedSynch_filling : 3.39005e-05 sec
OP_ID736. Dist_c1_fusedSynch_filling : 3.10952e-05 sec
OP_ID740. Dist_c1_fusedSynch_filling : 4.10645e-05 sec
OP_ID742. Dist_c1_fusedSynch_filling : 4.67419e-05 sec
OP_ID746. Dist_c1_fusedSynch_filling : 5.73343e-05 sec
OP_ID750. Dist_c1_fusedSynch_filling : 4.29597e-05 sec
OP_ID752. Dist_c1_fusedSynch_filling : 5.23285e-05 sec
OP_ID756. Dist_c1_fusedSynch_filling : 7.19762e-05 sec
OP_ID760. Dist_c1_fusedSynch_filling : 7.28232e-05 sec
OP_ID762. Dist_c1_fusedSynch_filling : 5.33092e-05 sec
OP_ID767. Dist_c1_fusedSynch_filling : 4.32056e-05 sec
OP_ID771. Dist_c1_fusedSynch_filling : 6.98061e-05 sec
OP_ID773. Dist_c1_fusedSynch_filling : 5.39216e-05 sec
OP_ID777. Dist_c1_fusedSynch_filling : 4.90921e-05 sec
OP_ID780. Dist_c1_fusedSynch_filling : 5.80866e-05 sec
OP_ID782. Dist_c1_fusedSynch_filling : 4.03034e-05 sec
OP_ID786. Dist_c1_fusedSynch_filling : 3.60825e-05 sec
OP_ID790. Dist_c1_fusedSynch_filling : 3.7162e-05 sec
OP_ID792. Dist_c1_fusedSynch_filling : 2.67683e-05 sec
OP_ID796. Dist_c1_fusedSynch_filling : 5.23501e-05 sec
OP_ID800. Dist_c1_fusedSynch_filling : 2.80245e-05 sec
OP_ID802. Dist_c1_fusedSynch_filling : 6.04765e-05 sec
OP_ID807. Dist_c1_fusedSynch_filling : 0.00010919 sec
OP_ID812. Dist_c1_fusedSynch_filling : 3.43743e-05 sec
OP_ID814. Dist_c1_fusedSynch_filling : 0.00071666 sec
OP_ID818. Dist_c1_fusedSynch_filling : 1.20282e-05 sec
OP_ID820. Dist_s_fusedSynch_allreduce : 0.00549486 sec
OP_ID822. Dist_c1_fusedSynch_copyBackToTensor : 0.00056166 sec
```

chrishkchris commented:
Ready for Review/merge

XJDKC commented on src/core/device/cuda_gpu.cc:
```diff
@@ -120,9 +132,41 @@ void CudaGPU::TimeProfilingDoExec(function<void(Context*)>&& fn, int executor,
   cudaEventCreate(&(node->start_));
   cudaEventCreate(&(node->end_));
 
+#ifdef USE_DIST
```
Do we have a better scheme?

chrishkchris replied:
If I use USE_CUDA instead if USE_DIST in common.h, this @ifdef can be deleted

chrishkchris replied:
which one do you prefer? I can use USE_CUDA in common.h instead

XJDKC replied:
I mean not to record the event based on the operator name.

chrishkchris replied:
then may pass the cuda stream from the buffered op and save the cuda stream in the node?

XJDKC replied:
Yeah, we need big changes to solve this problem(e.g. passing a struct named OpDescriptor to the device). I think we just keep it like this for now.

chrishkchris replied:
> Yeah, we need big changes to solve this problem(e.g. passing a struct named OpDescriptor to the device). I think we just keep it like this for now.

Yes, I guess OpDescriptor would be useful to record many context regarding the operator. I think it is good to include it in the future

XJDKC commented on src/io/communicator.cc:
```diff
@@ -179,14 +175,14 @@ void Communicator::wait() {
+        CUDA_CHECK(cudaStreamWaitEvent(ctx->stream, event, 0));
       },
-      blocks_, blocks_, "wait");
+      blocks_, blocks_, "Sync");
```
The operator name is a bit confusing.

chrishkchris replied:
I would like to rename all the waiting operators to be "waiting" so I can use a single condition in the if statement:
https://github.com/chrishkchris/singa/blob/SINGA-510_2/src/core/scheduler/scheduler.cc#L292

For the typo, I need to change it back to waiting instead of wait

XJDKC replied:
> I would like to rename all the waiting operators to be "waiting" so I can use a single condition in the if statement:
> https://github.com/chrishkchris/singa/blob/SINGA-510_2/src/core/scheduler/scheduler.cc#L292
> 
> For the typo, I need to change it back to waiting instead of wait

Got it.

XJDKC replied:
Got it.

XJDKC commented:
@chrishkchris In the communicator.cc, we split the original operation into multiple. Is this to facilitate recording the time?

chrishkchris commented:
> @chrishkchris In the communicator.cc, we split the original operation into multiple. Is this to facilitate recording the time?

Yes, in order to record the time, those operations that are used for waiting should not be included.  Otherwise the recorded time will include the waiting time.

For example, if I do not split the operator, a filling buffer operation will have the time profiled as the sum of :
1. waiting for the NULL stream before start filling buffer 2. filling buffer
If there is no need to wait, then the time spent for 1 will be zero, but some it still need to wait. 

The aims of the time profiling design is to figure out how much time actual time (excluding waiting time) we need for each operators, so in the future we may be able to do better scheduling.

Therefore, all the "waiting" operation should be excluded from time profiling of operators.

Also, since there is no need to record the waiting time, I renamed all the waiting operators as "waiting" so that we don't time profiling for them
https://github.com/chrishkchris/singa/blob/SINGA-510_2/src/core/scheduler/scheduler.cc#L292

chrishkchris commented on src/io/communicator.cc:
```diff
@@ -134,7 +131,6 @@ void Communicator::sparsInit() {
   CUDA_CHECK(cudaMalloc(&xInd, (int)(sizeof(int) * maxSize)));
   CUDA_CHECK(cudaMalloc(&xVal, (int)(sizeof(float) * maxSize)));
   CUSPARSE_CHECK(cusparseCreate(&cusparse_handle));
-  CUSPARSE_CHECK(cusparseSetStream(cusparse_handle, c2));
```
Yes, now the stream is not in the communicator.
Even in the orginal version, I was using different streams for different operations in cusparse.
The API let me stream for cusparse right before calling the cusparse operators, it let the users to use different streams for different operations

XJDKC replied:
I see, you moved this line to the position before each call to sparse.

chrishkchris replied:
In the original kernal I also did in this way of many operators

chrishkchris commented on include/singa/io/communicator.h:
```diff
@@ -98,16 +98,16 @@ class Communicator {
   void generateBlocks(std::vector<Tensor> &t);
   void allReduce(int size, void *sendbuff, void *recvbuff,
-                 ncclDataType_t ncclType);
+                 ncclDataType_t ncclType, Context *ctx);
```
oh I will go to delete the three lines in the communicator.h header:
  cudaStream_t s;
  cudaStream_t c1;
  cudaStream_t c2;

XJDKC commented:
You are right. The waiting time cannot be included in the execution time of the operation. But for some operators that use two cuda streams, we determine which stream to record events based on the name of the operator. I think it's not an elegant scheme, any ideas about this?

chrishkchris commented:
> You are right. The waiting time cannot be included in the execution time of the operation. But for some operators that use two cuda streams, we determine which stream to record events based on the name of the operator. I think it's not an elegant scheme, any ideas about this?

For time profiling, the idea situation is: All the buffered communicator operators should use only one cuda stream, two streams is not good because one stream should wait for another. So I broke down most of the operations.

The only one kernal I did not yet break it down yet is the sparse communication kernal, which is too long so I do not inlcude breaking the kernal down in this PR.
https://github.com/chrishkchris/singa/blob/SINGA-510_2/src/io/communicator.cc#L444

My original plan of this PR is record the fp32/fp16 communication time seamlessly. If it prodives better time profiling for the sparse communication (breaking the large kernal down), it can be included in the future PR

chrishkchris commented on include/singa/core/common.h:
```diff
@@ -100,13 +100,21 @@ typedef struct _Context {
 #endif
 #endif  // USE_CUDA
 
+#ifdef USE_DIST
```
In the CPU only version, there is no cuda library. It will not have those cuda related headers.
In the building process, the default USE_DIST is 0.
So the Travis CI build of CPU version didn't include this part, you may check the CI CPU version build result
My concern not to use USE_CUDA instead is that it could add three more useless cuda stream into the single GPU version.

chrishkchris replied:
All of them in Travis CI build are only CPU version
https://travis-ci.org/github/apache/singa/builds/693742421

XJDKC commented:
Summary:
* fix typos
* remove some outdated code mentioned above.
Future plan:
* maybe we can add a struct named OpDescriptor to record ever-increasing context regarding the operator
* the time of the same type of operator can be accumulated and sorted, which is convenient for viewing the execution time of a certain type of operator. Then we can know where the bottleneck is.

chrishkchris commented:
> Summary:
> 
> * fix typos
> * remove some outdated code mentioned above.
> 
> Future plan:
> 
> * maybe we can add a struct named OpDescriptor to record ever-increasing context regarding the operator
> * the time of the same type of operator can be accumulated and sorted, which is convenient for viewing the execution time of a certain type of operator. Then we can know where the bottleneck is.

thanks for the summary!
